spring-kafka/src/test/java/org/springframework/kafka/listener/KafkaMessageListenerContainerTests.java

```diff
@@ -560,6 +565,161 @@ public void testBatchAck() throws Exception {
+		KafkaTemplate<Integer, String> template = new KafkaTemplate<>(pf);
+		template.setDefaultTopic(topic9);
+		template.sendDefault(0, 0, "foo");
+		template.sendDefault(0, 0, "baz");
+		template.sendDefault(1, 0, "bar");
+		template.sendDefault(1, 0, "qux");
+		template.flush();
+
+		Map<String, Object> props = KafkaTestUtils.consumerProps("test9", "false", embeddedKafka);
+		props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+		DefaultKafkaConsumerFactory<Integer, String> cf = new DefaultKafkaConsumerFactory<>(props);
+		ContainerProperties containerProps = new ContainerProperties(topic9);
+		final CountDownLatch latch = new CountDownLatch(4);
+		containerProps.setMessageListener((BatchAcknowledgingMessageListener<Integer, String>) (messages, ack) -> {
+			logger.info("batch listener manual: " + messages);
+			for (int i = 0; i < messages.size(); i++) {
+				latch.countDown();
+			}
+			ack.acknowledge();
+		});
+		containerProps.setSyncCommits(true);
+		containerProps.setAckMode(AckMode.MANUAL_IMMEDIATE);
+		containerProps.setPollTimeout(10000);
+		containerProps.setAckOnError(false);
+
+		KafkaMessageListenerContainer<Integer, String> container = new KafkaMessageListenerContainer<>(cf,
+				containerProps);
+		container.setBeanName("testBatchListenerManual");
+		container.start();
+
+		assertThat(latch.await(60, TimeUnit.SECONDS)).isTrue();
+		Consumer<Integer, String> consumer = cf.createConsumer();
+		consumer.assign(Arrays.asList(new TopicPartition(topic9, 0), new TopicPartition(topic9, 1)));
+		assertThat(consumer.position(new TopicPartition(topic9, 0))).isEqualTo(2);
+		assertThat(consumer.position(new TopicPartition(topic9, 1))).isEqualTo(2);
+		container.stop();
+		consumer.close();
+		logger.info("Stop batch listener manual");
+	}
+
+	@Test
+	public void testBatchListenerErrors() throws Exception {
+		logger.info("Start batch listener errors");
+
+		Map<String, Object> senderProps = KafkaTestUtils.producerProps(embeddedKafka);
+		ProducerFactory<Integer, String> pf = new DefaultKafkaProducerFactory<>(senderProps);
+		KafkaTemplate<Integer, String> template = new KafkaTemplate<>(pf);
+		template.setDefaultTopic(topic10);
+		template.sendDefault(0, 0, "foo");
+		template.sendDefault(0, 0, "baz");
+		template.sendDefault(1, 0, "bar");
+		template.sendDefault(1, 0, "qux");
+		template.flush();
+
+		Map<String, Object> props = KafkaTestUtils.consumerProps("test9", "false", embeddedKafka);
+		props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+		DefaultKafkaConsumerFactory<Integer, String> cf = new DefaultKafkaConsumerFactory<>(props);
+		ContainerProperties containerProps = new ContainerProperties(topic10);
+		containerProps.setMessageListener((BatchMessageListener<Integer, String>) messages -> {
+			logger.info("batch listener errors: " + messages);
+			throw new RuntimeException("intentional");
+		});
+		containerProps.setSyncCommits(true);
```

Comment:
Well, I think we still have some race condition with such an approach.
Yes, the commit is sync, indeed. But it is done in the `ListenerConsumer` Thread, not our main to let us to be blocked until commit.
Therefore I think all those assertions in the end are going to fail from time to time.